openid-4-verifiable-presentations-1_0.md

```diff
@@ -616,11 +622,310 @@ The Wallet then validates the request as specified in OAuth 2.0 [@RFC6749].
+
+### Selecting Claims and Credentials {#vp_query_lang_processing_rules}
+
+The following section describes the logic that applies for selecting claims 
+and for selecting credentials. 
+
+Note: While this specification provides the mechanisms for requesting different
+sets of claims and credentials, it does not make assumptions about the user 
+interface of the Wallet, for example, if users can select which claims to present 
+or which combination of credentials to present.
+
+#### Selecting Claims {#selecting_claims}
+
+The following rules apply for selecting claims via `claims` and `claim_sets`:
+
+- If `claims` is not provided, the Verifier requests all claims existing
+  in the Credential.
+- If `claims` is provided, but `claim_sets` is not provided,
+  the Verifier requests all claims listed in `claims`.
+- Otherwise, the Verifier requests one combination of the claims listed in
+  `claim_sets`. The order of the options conveyed in the `claim_sets`
+  array expresses the Verifiers' preference for what is returned; the Wallet MUST return
+  the first option that it can satisfy. If the Wallet cannot satisfy any of the
+  options, it MUST NOT return any claims.
+
+When a Claims Query contains a restriction on the values of a claim, the Wallet
+SHOULD NOT return the claim if its value does not match at least one of the
+elements in `values` i.e., the claim should be treated the same as if it did not
+exist in the Credential. Implementing this restriction may not be possible in
+all cases, for example, if the Wallet does not have access to the claim value
+before presentation or user consent or if another component routing
+the request to the Wallet does not have access to the claim value. Therefore, Verifiers
+must treat restrictions expressed using `values` as a best-effort way to improve
+user privacy, but MUST NOT rely on it for security checks.
+
+The purpose of the `claim_sets` syntax is to provide a way for a 
+verifier to describe alternative ways a given credential can satisfy the request. 
+The array ordering expresses the Verifier's preference for how the request must 
+be fulfilled. The first element in the array is the most preferred and the last 
+element in the array is the least preferred. Verifiers SHOULD use the principle of 
+least information disclosure to influence how they order these options. For example, a 
+proof of age request should prioritize requesting an attribute like `age_over_18` over
+an attribute like `birth_date`.
+
+If the Wallet cannot deliver all claims requested by the Verifier
+according to these rules, it MUST NOT return the respective Credential.
+
+#### Selecting Credentials
+
+The following rules apply for selecting Credentials via `credentials` and `credential_sets`:
+
+- If `credential_sets` is not provided, the Verifier requests all
+  Credentials in `credentials` to be returned.
+- Otherwise, the Verifier requests all of the credential set queries in the `credential_sets` array
```

Comment:
```suggestion
- Otherwise, if `credential_sets` is provided, and  the `required` property is set to true or is absent,  the Verifier is requesting that all of the credentials from one of  the credential sets in the "query_options" is returned. 
```

Reply:
I prefer @danielfett language here, this suggested language is much harder to grok.

Reply:
Agree with @tplooker. Also, the proposed wording doesn't quite work — `required` is defined per entry in `credential_sets`.

Reply:
But the original wording is ambiguous at best, and not understandable at worst. So more clear wording is definitely needed.

Reply:
The whole text from line 791 to 793

> Otherwise, the Verifier requests all of the credential set queries in the `credential_sets` array with `required` evaluated as true to be returned at a minimum and optionally any of the credential set queries with `required` evaluated as false.